### PR TITLE
feat(httpd): implement install and remove enpoints

### PR DESCRIPTION
**Pull request by tareknaser360**
This pull request adds two new endpoints:

-  `coffee_install`  
- `coffee_remove` 

To test the new endpoints: 
- 
-
```
curl -X 'POST' \
  'http://localhost:8080/install' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "plugin": "summary"
}' 
```
- 
```
curl -X 'POST' \
  'http://localhost:8080/remove' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "plugin": "summary"
}' 
```

### CODE REVIEW

File: coffee_httpd/src/httpd/server.rs [hunk 1]
```diff
@@ -73,6 +75,48 @@ async fn coffee_help(
     Ok(body)
 }
 
+#[api_v2_operation]
+#[post("/install")]
+async fn coffee_install(
+    data: web::Data<AppState>,
+    body: Json<HashMap<String, String>>,
```
Comment (vincenzopalazzo) [inline coffee_httpd/src/httpd/server.rs:82]:
```suggestion
    body: Json<InstallBody>,
```

I guess you can do something similar and avoid to bring the hash map around

File: coffee_httpd/src/httpd/server.rs [hunk 2]
```diff
@@ -73,6 +75,48 @@ async fn coffee_help(
...
+            "Failed to install plugin: {err}"
+        ))),
+    }
+}
+
+#[api_v2_operation]
+#[post("/remove")]
+async fn coffee_remove(
+    data: web::Data<AppState>,
+    body: Json<HashMap<String, String>>,
```
Comment (vincenzopalazzo) [inline coffee_httpd/src/httpd/server.rs:103]: same here

File: coffee_httpd/src/httpd/server.rs [hunk 3]
```diff
@@ -73,6 +75,48 @@ async fn coffee_help(
...
+) -> Result<HttpResponse, Error> {
+    let plugin = body.get("plugin").ok_or_else(|| {
+        actix_web::error::ErrorBadRequest("Missing 'plugin' field in the request body")
+    })?;
+
+    let mut coffee = data.coffee.lock().await;
+    let result = coffee.remove(plugin).await;
+
+    match result {
+        Ok(_) => Ok(HttpResponse::Ok().body(format!("Plugin '{}' removed successfully", plugin))),
```
Comment (vincenzopalazzo) [inline coffee_httpd/src/httpd/server.rs:113]:
```suggestion
        Ok(_) => Ok(HttpResponse::Ok().body(format!("Plugin '{}' removed successfully", plugin))),
```

Can we log the value that it is inside the `Ok(_)`?

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Some comments but overall looks good

File: coffee_httpd/src/httpd/server.rs (hunk 1 above)
Comment (tareknaser360) [reply coffee_httpd/src/httpd/server.rs]: Yes. I fixed this also for the previously implemented endpoints that accept a request body.
Now for testing, we can 
```
curl -X POST \
  http://localhost:8080/remote/rm \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
    "repository_name": "Folgore"
}'
```

File: coffee_httpd/src/httpd/server.rs [hunk 4]
```diff
@@ -13,6 +13,7 @@ use std::sync::Arc;
 use serde_json::Value;
 use tokio::sync::Mutex;
 
+use super::utils::*;
 use coffee_core::coffee::CoffeeManager;
 use coffee_lib::plugin_manager::PluginManager;
```
Comment (vincenzopalazzo) [inline coffee_httpd/src/httpd/server.rs:18]:
```suggestion
use coffee_core::coffee::CoffeeManager;
use coffee_lib::plugin_manager::PluginManager;

use super::utils::*;
```

I need to write down some rules for the import for now we move down it

File: coffee_httpd/src/httpd/server.rs [hunk 5]
```diff
@@ -51,6 +52,9 @@ pub async fn run_httpd<T: ToSocketAddrs>(
             .wrap_api()
             .service(swagger_api)
             .service(coffee_help)
+            .service(coffee_setup)
```
Comment (vincenzopalazzo) [inline coffee_httpd/src/httpd/server.rs:55]:
```suggestion
            .service(coffee_setup)
```

I don't think this makes sense. The endpoint will be authenticated in some way so I guess the httpd makes sense only when core lightning is already started.

imho in the case of the httpd server needs to know the path of core lightning as an argument to the `coffee_httpd --cln-path=....` 

I do not have any use case in mind to support the setup endpoint

File: coffee_httpd/src/httpd/server.rs [hunk 6]
```diff
@@ -74,6 +78,67 @@ async fn coffee_help(
...
+    let mut coffee = data.coffee.lock().await;
+    let result = coffee.setup(cln_dir).await;
+
+    match result {
+        Ok(_) => Ok(HttpResponse::Ok().body(format!("Coffee setup completed successfully"))),
+        Err(err) => Err(actix_web::error::ErrorInternalServerError(format!(
+            "Failed to perform Coffee setup: {err}"
+        ))),
+    }
+}
```
Comment (vincenzopalazzo) [inline coffee_httpd/src/httpd/server.rs:98]: ok this can be removed

File: coffee_httpd/src/httpd/server.rs [hunk 7]
```diff
@@ -74,6 +78,67 @@ async fn coffee_help(
...
+
+    let mut coffee = data.coffee.lock().await;
+    let result = coffee.install(plugin, false, try_dynamic).await;
+
+    match result {
+        Ok(_) => Ok(HttpResponse::Ok().body(format!("Plugin '{}' installed successfully", plugin))),
+        Err(err) => Err(actix_web::error::ErrorInternalServerError(format!(
+            "Failed to install plugin: {err}"
+        ))),
+    }
```
Comment (vincenzopalazzo) [inline coffee_httpd/src/httpd/server.rs:117]: we should write a rust macro to avoid always write this code

File: coffee_httpd/src/httpd/utils.rs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+#[derive(Debug, Deserialize, Apiv2Schema, Serialize)]
+pub struct RemoteAddRequest {
+    pub repository_name: String,
+    pub repository_url: String,
+}
+
+#[derive(Debug, Deserialize, Apiv2Schema, Serialize)]
+pub struct RemoteRmRequest {
+    pub repository_name: String,
+}
```
Comment (vincenzopalazzo) [inline coffee_httpd/src/httpd/utils.rs:29]:
```suggestion
pub mod request {
#[derive(Debug, Deserialize, Apiv2Schema, Serialize)]
pub struct SetupRequest {
    pub cln_dir: String,
}

#[derive(Debug, Deserialize, Apiv2Schema, Serialize)]
pub struct InstallRequest {
    pub plugin: String,
    pub try_dynamic: bool,
}

#[derive(Debug, Deserialize, Apiv2Schema, Serialize)]
pub struct RemoveRequest {
    pub plugin: String,
}

#[derive(Debug, Deserialize, Apiv2Schema, Serialize)]
pub struct RemoteAddRequest {
    pub repository_name: String,
    pub repository_url: String,
}

#[derive(Debug, Deserialize, Apiv2Schema, Serialize)]
pub struct RemoteRmRequest {
    pub repository_name: String,
}
}
```

and then remove the suffix Request from the name of the struct

then you have `use utils::request;` and use use the struct as `request::AddRemote`

File: tests/Cargo.toml [hunk 9]
```diff
@@ -6,6 +6,7 @@ edition = "2021"
 [dependencies]
 coffee_lib = { path = "../coffee_lib" }
 coffee_testing = { path = "../coffee_testing" }
+coffee_httpd = { path = "../coffee_httpd" }
```
Comment (vincenzopalazzo) [inline tests/Cargo.toml:9]: Mh, I am thinking, as model we can use what we have already inside the coffee_lib? https://github.com/coffee-tools/coffee/blob/master/coffee_lib/src/types/mod.rs

Maybe make two modules as `request` and `response`?

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
OK the CI is happy and we are all happy too

I added a couple of suggestion below

File: coffee_httpd/src/main.rs [hunk 10]
```diff
@@ -14,7 +14,8 @@ async fn main() {
     if let Err(err) = &coffee {
         println!("{err}");
     }
-    let coffee = coffee.unwrap();
+    let mut coffee = coffee.unwrap();
+    coffee.setup(&cmd.cln_path).await.unwrap();
```
Comment (vincenzopalazzo) [inline coffee_httpd/src/main.rs:18]: we should remove this unwrap from here

File: coffee_lib/src/macros.rs [hunk 11]
```diff
@@ -48,5 +48,18 @@ macro_rules! sh {
...
+    ($result:expr, $msg:expr) => {
+        match $result {
+            Ok(_) => Ok(HttpResponse::Ok().body(format!($msg))),
+            Err(err) => Err(actix_web::error::ErrorInternalServerError(format!(
+                "Error: {}",
+                err
+            ))),
+        }
+    };
+}
```
Comment (vincenzopalazzo) [inline coffee_lib/src/macros.rs:63]: no this should stay in the httpd, we do not have access to the HTTP Response here

File: coffee_lib/src/types/mod.rs [hunk 12]
```diff
@@ -1,58 +1,95 @@
...
+    use serde::{Deserialize, Serialize};
 
-#[derive(Debug, Serialize, Deserialize)]
-pub struct CoffeeRemove {
-    pub plugin: Plugin,
-}
+    #[derive(Debug, Deserialize, Apiv2Schema, Serialize)]
+    pub struct Setup {
+        pub cln_dir: String,
+    }
```
Comment (vincenzopalazzo) [inline coffee_lib/src/types/mod.rs:11]: we do not have this endpoint so we can remove it

File: coffee_testing/src/lib.rs [hunk 13]
```diff
@@ -42,7 +42,7 @@ pub mod macros {
 
     #[macro_export]
     macro_rules! httpd {
-        ($dir:expr, $port:expr, $($opt_args:tt)*) => {
+        ($dir:expr, $port:expr, $cln_path:expr, $($opt_args:tt)*) => {
```
Comment (vincenzopalazzo) [inline coffee_testing/src/lib.rs:45]: no need to had this here, you can add this to the macros string, see next suggestion

File: coffee_testing/src/lib.rs [hunk 14]
```diff
@@ -172,22 +174,29 @@ impl Drop for CoffeeHTTPDTesting {
 
 impl CoffeeHTTPDTesting {
     /// init coffee httpd in a tmp directory.
-    pub async fn tmp() -> anyhow::Result<Self> {
+    pub async fn tmp(cln_path: String) -> anyhow::Result<Self> {
         let dir = tempfile::tempdir()?;
         let port = port::random_free_port().unwrap();
-        let child = httpd!(dir, port, "{}", format!("--network=regtest"))?;
+        let child = httpd!(dir, port, cln_path, "{}", format!("--network=regtest"))?;
```
Comment (vincenzopalazzo) [inline coffee_testing/src/lib.rs:180]:
```suggestion
        let child = httpd!(dir, port,  "{}", format!("--network=regtest --cln-path={cln_path}"))?;
```

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Ok I made another pass to this, but we are close

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]

File: coffee_httpd/src/main.rs [hunk 15]
```diff
@@ -15,7 +15,9 @@ async fn main() {
         println!("{err}");
     }
     let mut coffee = coffee.unwrap();
-    coffee.setup(&cmd.cln_path).await.unwrap();
+    if let Err(err) = coffee.setup(&cmd.cln_path).await {
+        println!("{err}");
+    }
```
Comment (vincenzopalazzo) [inline coffee_httpd/src/main.rs:20]: do you think that we should return the error to the caller? like `async main() -> Result<()>`

File: coffee_lib/Cargo.toml [hunk 16]
```diff
@@ -12,3 +12,6 @@ log = "0.4.17"
 env_logger = "0.9.3"
 tokio = { version = "1.22.0", features = ["process"] }
 paperclip = { version = "0.8.0", features = ["actix4"] }
+
+[features]
+httpd = ["paperclip/actix4"]
```
Comment (vincenzopalazzo) [inline coffee_lib/Cargo.toml:17]:
```suggestion
httpd = ["dep:paperclip"]
```

File: coffee_lib/Cargo.toml [hunk 17]
```diff
@@ -12,3 +12,6 @@ log = "0.4.17"
 env_logger = "0.9.3"
 tokio = { version = "1.22.0", features = ["process"] }
 paperclip = { version = "0.8.0", features = ["actix4"] }
```
Comment (vincenzopalazzo) [inline coffee_lib/Cargo.toml:14]:
```suggestion
paperclip = { version = "0.8.0", features = ["actix4"], optional = true }
```

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]

**Review by vincenzopalazzo — APPROVED** [review]

File: coffee_lib/src/types/mod.rs [hunk 18]
```diff
@@ -1,58 +1,90 @@
...
+    pub struct Install {
+        pub plugin: String,
+        pub try_dynamic: bool,
+    }
 
-#[derive(Debug, Serialize, Deserialize)]
-pub struct CoffeeList {
-    pub plugins: Vec<Plugin>,
-}
+    #[derive(Debug, Deserialize, Apiv2Schema, Serialize)]
```
Comment (vincenzopalazzo) [inline coffee_lib/src/types/mod.rs:14]:
```suggestion
     #[cfg(open-api)]
    #[derive(Apiv2Schema)]
    #[derive(Debug, Deserialize, Apiv2Schema, Serialize)]
```

I do not know if this works but `Apiv2Schema` is not present always

File: coffee_lib/src/types/mod.rs [hunk 19]
```diff
@@ -1,58 +1,90 @@
 //! Coffee Model Definition
-use serde::{Deserialize, Serialize};
 
-use crate::plugin::Plugin;
+// Definition of the request types.
+pub mod request {
+    use paperclip::actix::Apiv2Schema;
```
Comment (vincenzopalazzo) [inline coffee_lib/src/types/mod.rs:5]:
```suggestion
    #[cfg(open-api)]
    use paperclip::actix::Apiv2Schema;
```

this is not present aways

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]

**Review by vincenzopalazzo — APPROVED** [review]

**Pull request merged**